presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -0,0 +1,289 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.airlift.log.Logger;
+import com.facebook.presto.sql.tree.AllColumns;
+import com.facebook.presto.sql.tree.ArithmeticBinaryExpression;
+import com.facebook.presto.sql.tree.AstVisitor;
+import com.facebook.presto.sql.tree.ComparisonExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.GroupBy;
+import com.facebook.presto.sql.tree.GroupingElement;
+import com.facebook.presto.sql.tree.Identifier;
+import com.facebook.presto.sql.tree.LogicalBinaryExpression;
+import com.facebook.presto.sql.tree.Node;
+import com.facebook.presto.sql.tree.OrderBy;
+import com.facebook.presto.sql.tree.Query;
+import com.facebook.presto.sql.tree.QueryBody;
+import com.facebook.presto.sql.tree.QuerySpecification;
+import com.facebook.presto.sql.tree.Relation;
+import com.facebook.presto.sql.tree.Select;
+import com.facebook.presto.sql.tree.SelectItem;
+import com.facebook.presto.sql.tree.SimpleGroupBy;
+import com.facebook.presto.sql.tree.SingleColumn;
+import com.facebook.presto.sql.tree.SortItem;
+import com.facebook.presto.sql.tree.Table;
+import com.facebook.presto.sql.tree.TableSubquery;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class MaterializedViewQueryOptimizer
+        extends AstVisitor<Node, Void>
+{
+    private static final Logger logger = Logger.get(MaterializedViewQueryOptimizer.class);
+    private final Table materializedView;
+    private final ImmutableMap<String, String> baseToViewColumnMap;
+    private final Optional<Relation> baseQueryRelation;
+
+    public MaterializedViewQueryOptimizer(Table materializedView, Query originalSqlQuery)
+    {
+        this.materializedView = requireNonNull(materializedView, "materialized view is null");
+
+        QuerySpecification materializedViewDefinition = (QuerySpecification) originalSqlQuery.getQueryBody();
+        this.baseQueryRelation = materializedViewDefinition.getFrom().isPresent() ? materializedViewDefinition.getFrom() : Optional.empty();
+
+        // Create base to materialized view column map
+        ImmutableMap.Builder<String, String> baseToViewColumnMap = ImmutableMap.builder();
+
+        for (SelectItem baseTableSelectColumn : materializedViewDefinition.getSelect().getSelectItems()) {
+            if (baseTableSelectColumn instanceof SingleColumn) {
+                String baseColumnName = ((SingleColumn) baseTableSelectColumn).getExpression().toString();
+                Optional<Identifier> baseColumnAlias = ((SingleColumn) baseTableSelectColumn).getAlias();
+                String viewDerivedColumnName = baseColumnAlias.orElse(new Identifier(baseColumnName)).getValue();
```

Comment:
This is not safe:
- what if you have `max(a * b) as max1, max(a * b) as max2` in your query?
- what if you have `a * b` in your base view and your select query is `b * a`. The logic will miss that.
- what if you have WHERE clause in your select query or view query

Have a mapping of expression to select columns won't be enough. Let's have some thorough thinking into basic scan, filter, project, and aggregation query. We need to make sure the mv definition query is a proper subquery of the select query.